ocaml/typing/typecore.ml

```diff
@@ -382,30 +374,43 @@ let mode_default mode =
     exact = false;
     tuple_modes = [] }
 
-(* used when entering a function;
-mode is the mode of the function region *)
+(** used when entering a function; mode is the mode of the function region *)
 let mode_return mode =
   { (mode_default (Value_mode.local_to_regional mode)) with
     position = RTail (mode, FTail);
     escaping_context = Some Return;
   }
 
-(* used when entering a region.*)
-let mode_region mode =
-  { (mode_default (Value_mode.local_to_regional mode)) with
-    position = RTail (mode, FNontail);
-    escaping_context = None;
+
+(** We expect nothing for the modes, usually for subexpressions whose type must
+  cross modes. *)
+let mode_max = mode_default Value_mode.max_mode
+
+(** Same as [mode_max] but set [position] *)
+let mode_max_with_position position = { mode_max with position }
+
+(** used when entering a region whose type must cross modes *)
+let mode_max_with_region =
+  { (mode_default Value_mode.regional) with
+    position = RTail (Value_mode.local, FNontail);
   }
 
-let mode_local =
-  mode_default Value_mode.local
+(** Used for several cases; all non-tail *)
+let mode_exact mode = { (mode_default mode) with exact = true }
```

Comment:
This one doesn't have an obvious semantic meaning, I think it would be better to have this one be used only for the uncurried function and recursive functions case, where we need to force some functions to have exactly the same mode because of how we compile them. Then inline this into the `mode_local` function below, and use that for the exclave case too.